server/src/main/java/org/opensearch/rest/pagination/ShardPaginationStrategy.java

```diff
@@ -0,0 +1,244 @@
+            ? null
+            : new ShardStrategyToken(requestedTokenStr);
+        if (Objects.isNull(requestedToken)) {
+            return indexMetadata -> true;
+        }
+        return metadata -> {
+            if (metadata.getIndex().getName().equals(requestedToken.lastIndexName)) {
+                return true;
+            } else if (metadata.getCreationDate() == requestedToken.lastIndexCreationTime) {
+                return isAscendingSort
+                    ? metadata.getIndex().getName().compareTo(requestedToken.lastIndexName) > 0
+                    : metadata.getIndex().getName().compareTo(requestedToken.lastIndexName) < 0;
+            }
+            return isAscendingSort
+                ? metadata.getCreationDate() > requestedToken.lastIndexCreationTime
+                : metadata.getCreationDate() < requestedToken.lastIndexCreationTime;
+        };
+    }
+
+    private Tuple<List<ShardRouting>, List<IndexMetadata>> getPageData(
+        Map<String, IndexRoutingTable> indicesRouting,
+        List<IndexMetadata> sortedIndices,
+        final int pageSize,
+        String requestedTokenStr
+    ) {
+        List<ShardRouting> shardRoutings = new ArrayList<>();
+        List<IndexMetadata> indexMetadataList = new ArrayList<>();
+        ShardStrategyToken requestedToken = Objects.isNull(requestedTokenStr) || requestedTokenStr.isEmpty()
+            ? null
+            : new ShardStrategyToken(requestedTokenStr);
+        int shardCount = 0;
+        for (IndexMetadata indexMetadata : sortedIndices) {
+            boolean indexShardsAdded = false;
+            Map<Integer, IndexShardRoutingTable> indexShardRoutingTable = indicesRouting.get(indexMetadata.getIndex().getName())
+                .getShards();
+            int shardId = Objects.isNull(requestedToken) ? 0
+                : indexMetadata.getIndex().getName().equals(requestedToken.lastIndexName) ? requestedToken.lastShardId + 1
+                : 0;
+            for (; shardId < indexShardRoutingTable.size(); shardId++) {
+                shardCount += indexShardRoutingTable.get(shardId).size();
+                if (shardCount > pageSize) {
+                    break;
+                }
+                shardRoutings.addAll(indexShardRoutingTable.get(shardId).shards());
+                indexShardsAdded = true;
+            }
+            // Add index to the list if any of its shard was added to the count.
+            if (indexShardsAdded) {
+                indexMetadataList.add(indexMetadata);
+            }
+            if (shardCount >= pageSize) {
+                break;
+            }
+        }
+
+        return new Tuple<>(shardRoutings, indexMetadataList);
+    }
+
+    private PageToken getResponseToken(final int pageSize, final int totalIndices, IndexMetadata lastIndex, ShardRouting lastShard) {
+        if (totalIndices <= pageSize && lastIndex.getNumberOfShards() == lastShard.getId()) {
```

Comment:
isn't the pageSize here is number of shards ? Why is it compared with `totalIndices` ?